os/arch/arm/src/sidk_s5jt200/src/s5jt200_alc5658.c

```diff
@@ -0,0 +1,345 @@
+		g_alc5658info.handler = NULL;
+		g_alc5658info.arg = NULL;
+	}
+
+	return OK;
+}
+
+static bool alc5658_enable(FAR const struct alc5658_lower_s *lower, bool enable)
+{
+	static bool enabled;
+	irqstate_t flags;
+	bool ret;
+
+	/* Has the interrupt state changed */
+
+	flags = enter_critical_section();
+	if (enable != enabled) {
+		/* Enable or disable interrupts */
+
+		if (enable && g_alc5658info.handler) {
+			audinfo("Enabling\n");
+// DO IRQ ENABLING HERE
+			enabled = true;
+		} else {
+			audinfo("Disabling\n");
+// DO IRQ DISABLING HERE
+			enabled = false;
+		}
+	}
+
+	ret = enabled;
+	leave_critical_section(flags);
+	return ret;
+}
+
+static int alc5658_interrupt(int irq, FAR void *context)
+{
+	/* Just forward the interrupt to the ALC5658 driver */
+
+	audinfo("handler %p\n", g_alc5658info.handler);
+	if (g_alc5658info.handler) {
+		return g_alc5658info.handler(&g_alc5658info.lower, g_alc5658info.arg);
+	}
+
+	/* We got an interrupt with no handler.  This should not
+	 * happen.
+	 */
+
+	return OK;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: s5j_alc5658_initialize
+ *
+ * Description:
+ *   This function is called by platform-specific, setup logic to configure
+ *   and register the ALC5658 device.  This function will register the driver
+ *   as /dev/alc5658[x] where x is determined by the minor device number.
+ *
+ * Input Parameters:
+ *   minor - The input device minor number
+ *
+ * Returned Value:
+ *   Zero is returned on success.  Otherwise, a negated errno value is
+ *   returned to indicate the nature of the failure.
+ *
+ ****************************************************************************/
+int s5j_alc5658_initialize(int minor)
+{
+	FAR struct audio_lowerhalf_s *alc5658;
+	FAR struct audio_lowerhalf_s *pcm;
+	FAR struct i2c_dev_s *i2c;
+	FAR struct i2s_dev_s *i2s;
+	static bool initialized = false;
+	char devname[12];
+	int ret;
+
+	audinfo("minor %d\n", minor);
+	DEBUGASSERT(minor >= 0 && minor <= 25);
+
+	/* Have we already initialized?  Since we never uninitialize we must prevent
+	 * multiple initializations.  This is necessary, for example, when the
+	 * touchscreen example is used as a built-in application in NSH and can be
+	 * called numerous time.  It will attempt to initialize each time.
+	 */
+
+	if (!initialized) {
+		/* Configure the ALC5658 interrupt pin */
+
+		/* Do we have an interrupt PIN ? FIXME */
+
+		/* Get an instance of the I2C interface for the ALC5658 chip select */
+
+		i2c = up_i2cinitialize(CONFIG_ALC5658_I2C_PORT);
+		if (!i2c) {
+			auderr("ERROR: Failed to initialize CODEC I2C%d\n");
+			ret = -ENODEV;
+			goto errout;
+		}
+
+		/* Get an instance of the I2S interface for the ALC5658 data channel */
+
+		i2s = (struct i2s_dev_s *) s5j_i2s_initialize();
```

Comment:
Like up_i2cinitialize, i2s also needs interface.
Defineing i2s in common header, and each codec chip should implements interface.